kernel: Unify and simplify constructor support

keith-packard opened the pull request:
There are two styles of constructor support in wide use. They are signified by which ELF section the function addresses are placed in; the '.ctors' style and the '.init_array' style. Each constructor definition places the address of the function into a separate section using one of those two names (optionally appending a priority value).

The linker gathers these together and prepares data for the runtime to execute them in the correct order. Traditionally, the resulting .ctors section starts with the total count of constructors and ends with a terminating NULL while the .init_array section is simply a list of function pointers.

Keeping these separate was required when the linker and runtime were distributed independently; the runtime would look for both styles and execute them in sequence using slightly different code. A typical toolchain would only end up using one of the two styles, resulting in one of the two sections containing no function pointers. Zephyr had adopted this method as a fairly straightforward port of existing practice.

However, as Zephyr delivers both linker scripts and runtime in the same package, we can eliminate the redundancy and merge both .ctors and .init_array output from the compiler into a single list of initialization functions. This saves memory in ROM by eliminating the unnecessary .ctors count/NULL values, and saves code space by eliminating the .ctors handling.

Finally, all of this mechanism also applies to the ARC environment which had used only the .ctors sections but in an .init_array style.

See #74682

keith-packard committed:
kernel: Unify .ctors and .init_array handling

Handle both of these sections in a single chunk of code instead of
separately. We don't need to use the legacy .ctors ABI as both
the constructors array and startup logic are managed within a single
link result.

This can now also be used with ARC MWDT which had been using the .ctors
sections but with .init_array semantics. For ARC MWDT, we now always
discard .dtors and .fini sections as Zephyr will never cause global
destructors to execute. Stop discarding .eh_frame sections so that
exception handling works as expected.

When building a NATIVE_APPLICATION, we ask the native C library to run all
of the constructors to ensure any non-Zephyr constructors are run before
main is invoked. It might be "nice" to split the constructors so that the
Zephyr constructors were executed by the Zephyr code while the non-Zephyr
ones were executed by the native C library. I think that could be done if
we knew the pathnames of either the Zephyr or non-Zephyr files. That might
make a good future enhancement.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
tests/kernel: Validate constructor execution order

Check the order of constructors, making sure those with a specified
priority are called in numeric order after which those without priority are
called.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
Ok, I added a check in the linker script which will fail the build if the application wants to use constructors but the option to include the code to call them is disabled. With that, users can "safely" disable constructors, knowing that they won't accidentally end up needing this option.

keith-packard commented:
It looks like GNU-style constructors are not supported by `CMAKE_LINKER_GENERATOR`, so I've disabled constructors when that is enabled. Builds using C++ and the generator used to fail because the symbols required for the runtime code were not available. Now they will fail only if they use constructors due to the checks described above.

ceolin reviewed (commented):
Changes looks fine to me. Just wondering if this shouldn't be disabled by default.

keith-packard commented:
> Changes looks fine to me. Just wondering if this shouldn't be disabled by default.

That's a fine question. Now that the linker will generate an error if you need this option, we could safely disable it by default.

I think we could also make this "zero cost" by appending all of the constructors and initializers to the POST_KERNEL init level array in the linker. There's absolutely no reason to split these out, and that way we could remove a bunch of extra code and configuration bits. Let me try that.

aescolar commented on kernel/init.c:
```diff
@@ -433,10 +446,9 @@ static void bg_thread_main(void *unused1, void *unused2, void *unused3)
-	void z_cpp_init_static(void);
-	z_cpp_init_static();
-#endif /* CONFIG_CPP */
+#ifdef CONFIG_STATIC_INIT_GNU
```
This would be removing support for constructors for the arc mwdt toolchain.

keith-packard replied:
That seems bad. I'm really unclear as to how those constructors differ from the GNU constructors and whether we can support all of them in a single chunk of linker script.

evgeniy-paltsev replied:
@keith-packard 

For MWDT we just call internal `__do_global_ctors_aux` function via `z_cpp_init_static`:

https://github.com/zephyrproject-rtos/zephyr/blob/e5c05da9cbda7bc68aa84b796b828767b5eea6d7/lib/cpp/abi/cpp_init.c#L27

keith-packard replied:
So MWDT places all constructors in .ctors sections, just like some GCC architectures (most GCC targets now use .init_array for "reasons", I assume). I'm not seeing any incompatibility between MWDT and GCC here; we can just collect all of the .ctors and .init_array sections together into one giant array of initialization functions and call them in the right order. That should make everything the same, eliminating any MWDT specific bits.

The current v2/linker.ld script places the .ctors sections between `_fctors` and `_ectors`, which looks like it might be specific to some arc initialization sequence which is not known to Zephyr? Does that ever get used? I can't see how it could work...

keith-packard replied:
I've pushed a separate patch in this PR which removes the special case for MWDT toolchains; I'd appreciate help understanding if this is correct or not.

aescolar replied:
> The current v2/linker.ld script places the .ctors sections between _fctors and _ectors, which looks like it might be specific to some arc initialization sequence which is not known to Zephyr? 

I guess to whatever that toolchain provided `__do_global_ctors_aux()` does, which is not something I can see. But why not just leave the MWDT specific handling there by now? (we don't need to clean up everything in one go, specially if we risk breaking things we can't even try).

ruuddw replied:
The MetaWare `__do_global_ctors_aux` and `_fctors/_ectors` do something similar as the generic 'GNU' version, but it doesn't feel right to remove it. I'm not sure they indeed are fully equivalent, and if they are today, that may change in a next MetaWare version.
@evgeniy-paltsev, what do you think? Could we try and align/reuse GNU here, or should we stick with the MetaWare toolchain initialization?

keith-packard replied:
> But why not just leave the MWDT specific handling there by now? (we don't need to clean up everything in one go, specially if we risk breaking things we can't even try).

Yup, we figured that out yesterday :-). So, #74682 will be merged for 3.7 with just the fixes required to fix the bug.

keith-packard replied:
> The MetaWare `__do_global_ctors_aux` and `_fctors/_ectors` do something similar as the generic 'GNU' version, but it doesn't feel right to remove it. 

Unless `_fctors/_ectors` are used by code outside the Zephyr tree, then the current MWDT bits don't seem like they could possibly work? It feels like this was ported from some other system which used these symbols in the same way that Zephyr uses `__zephyr_init_array_start/__zephyr_init_array_end`.

MWDT *is* different from the generic GNU version currently though -- it uses `.ctors` sections but doesn't construct a "gnu compatible" ctors section. A GNU compatible ctors section consists of a count word followed by an array of function pointers and finally a NULL pointer.  The current MWDT ctors bits in Zephyr don't include the count or the trailing NULL.

This PR makes the GNU code match the MWDT code, removing the leading count and trailing NULL for the .ctors section.

> I'm not sure they indeed are fully equivalent, and if they are today, that may change in a next MetaWare version. @evgeniy-paltsev, what do you think? Could we try and align/reuse GNU here, or should we stick with the MetaWare toolchain initialization?

From what I can see, the MWDT compiler places all C++ constructor addresses in `.ctors` sections. If you use the Zephyr initialization code (which I think you must so that constructors can use `new`), then there isn't really any further MWDT dependency -- the Zephyr linker script defines how those addresses are laid out in memory, and then the Zephyr initialization code defines how that memory area is processed at runtime.

Reading through the ARC libgcc code leaves me with some questions though:

https://gcc.gnu.org/git/?p=gcc.git;a=blob;f=libgcc/config/arc/initfini.c;h=0183dd495ac5db165cc706ba2d62d9431e638972;hb=refs/heads/master

In that code, destructors are run in forward order while constructors are run in *reverse* order. All other systems run constructors in forward order (and destructors in reverse order).

aescolar reviewed (dismissed):
It's a bit of a weird situation to have a PR forked from another PR and both in review simultaneously.
Anyhow, this one as commented would remove support for CPP in the arc toolchain. Which i do not think is nice.
I anyhow took all your suggestions into the other PR.

keith-packard commented:
ok, fixed constructor priority problems (need to use `SORT_BY_INIT_PRIORITY` in the linker script). I also merged the ctors and init_array stuff -- different ABIs seem to use different section names, but the contents are compatible, so there's no reason to split them out. With that, the overhead for including GNU constructor support is down to 24 bytes on cortex-m0. I'd still consider disabling this all by default as the linker will throw an error if you need it.

keith-packard commented:
> It's a bit of a weird situation to have a PR forked from another PR and both in review simultaneously.

Yeah, we both pushed PRs at about the same time (I noticed yours while submitting mine). In any case, how about we figure out what code we want and get it into your PR and I'll close this one? The key changes I've made today are:

 * Merge ctors and init_array into a single list. There's no reason to have them split; the ctors section is used in some ABIs while init_array is used in others. The glibc ctors implementation isn't relevant to us as we have our own, so we don't need the count and null-termination entries. The result is a smaller implementation: 24 bytes on cortex-m0, down from 74 bytes.
 * Fix constructor priority handling -- you need to insert the constructors with priority value ahead of those without, and they need to be inserted in priority order using the linker `SORT_BY_INIT_PRIORITY` function.

aescolar commented:
> In any case, how about we figure out what code we want and get it into your PR and I'll close this one?

Let's do that. I'll take your latest changes into my PR tomorrow (its early night here already)

keith-packard commented:
I'll rebase this on top of main once #74682 is merged and leave it until after 3.7 is finished.

keith-packard commented:
I've recreated this PR as a simplification of the existing constructor implementation; this can wait to be merged after the release.

cfriedt commented:
This looks good to me. Is it already covered in a different branch

keith-packard commented:
> This looks good to me. Is it already covered in a different branch

I'm not sure which branch that would be -- #74682 fixed the bugs with this stuff; this patch is just a cleanup making things simpler -- merging .ctors/.init_array into a single form and unifying ARC MDWT support. The result is a bunch of code savings in both source and binary, but the semantics should be equivalent.

MaureenHelm commented:
@keith-packard coming back to this?

keith-packard commented:
Thanks for the reminder; I've been busy with picolibc recently and haven't been monitoring my zephyr PRs. I've rebased and pushed; it should 'just work' still; there weren't even any conflicts (!).

keith-packard commented:
Ok, this is ready for review now.

aescolar commented on include/zephyr/linker/common-rom/common-rom-init.ld:
```diff
@@ -1,76 +1,49 @@
 /* SPDX-License-Identifier: Apache-2.0 */
 
-#ifdef CONFIG_STATIC_INIT_GNU
```
(Continuing with the previous comment)
I think we should leave the conditions to choose this this as they are based on CONFIG_STATIC_INIT_GNU (together with the check)
```diff
diff --git a/Kconfig.zephyr b/Kconfig.zephyr
index 9f3e937f849..0e5c665b21d 100644
--- a/Kconfig.zephyr
+++ b/Kconfig.zephyr
@@ -434,7 +434,6 @@ config CODING_GUIDELINE_CHECK
 config NATIVE_LIBC
        bool
        select FULL_LIBC_SUPPORTED
-       select STATIC_INIT_GNU
        help
          Zephyr will use the host system C library.
 
diff --git a/include/zephyr/linker/common-rom/common-rom-init.ld b/include/zephyr/linker/common-rom/common-rom-init.ld
index 5523882a109..80587af1d1b 100644
--- a/include/zephyr/linker/common-rom/common-rom-init.ld
+++ b/include/zephyr/linker/common-rom/common-rom-init.ld
@@ -1,6 +1,6 @@
 /* SPDX-License-Identifier: Apache-2.0 */
 
-#ifdef CONFIG_TOOLCHAIN_SUPPORTS_STATIC_INIT_GNU
+#ifdef CONFIG_STATIC_INIT_GNU
        SECTION_PROLOGUE(init_array,,)
        {
                /*
@@ -14,36 +14,19 @@
                        SORT_BY_INIT_PRIORITY(.ctors.*)))
                KEEP (*(.init_array .ctors))
                __zephyr_init_array_end = .;
-#ifdef CONFIG_NATIVE_LIBC
-               /*
-                 * The __CTOR_LIST__ and __CTOR_END__ symbols are always defined
-                 * to result in an empty list. This is necessary to fix an issue
-                 * where the glibc process initialization code on native_posix
-                 * platforms calls constructors before Zephyr loads (issue #39347).
-                */
-               __CTOR_LIST__ = .;
-#ifdef CONFIG_64BIT
-               QUAD(0)
-               QUAD(0)
-#else
-               LONG(0)
-               LONG(0)
-#endif
-               __CTOR_END__ = .;
-#endif
-               /*
-               * Similar to the schenanigans required for the __CTOR_LIST__ and
-               * __CTOR_END__ symbols we define __init_array_start and __init_array_end
-               * to the same address to define an empty list. This prevents the glibc
-               * startup code from calling any global constructors before Zephyr loads.
-               */
-               __init_array_start = .;
-               __init_array_end = .;
        } GROUP_ROM_LINK_IN(RAMABLE_REGION, ROMABLE_REGION)
 
-#ifndef CONFIG_STATIC_INIT_GNU
-       ASSERT(__zephyr_init_array_start == __zephyr_init_array_end,
-              "GNU-style constructors required but STATIC_INIT_GNU not enabled")
-#endif
+#elif defined(CONFIG_TOOLCHAIN_SUPPORTS_STATIC_INIT_GNU) && !defined(CONFIG_NATIVE_APPLICATION)
+       /*
+        * If the code to invoke constructors is not enabled,
+        * make sure there aren't any in the application
+        */
+       SECTION_PROLOGUE(init_array,,)
+       {
+               KEEP(*(SORT_BY_NAME(".ctors*")))
+               KEEP(*(SORT_BY_NAME(".init_array*")))
+       } GROUP_ROM_LINK_IN(RAMABLE_REGION, ROMABLE_REGION)
 
+       ASSERT (SIZEOF(init_array) == 0,
+               "GNU-style constructors required but STATIC_INIT_GNU not enabled")
 #endif
``` 
Note I don't think we need anymore the code that creates the empty lists for native_posix
as with the current code I don't think we have anymore those problems (#39347 #36858)

keith-packard replied:
I think I need to switch from the Zephyr names to native C library names and then *always* include constructors when `NATIVE_LIBC` is enabled. That seems pretty simple.

keith-packard replied:
ok, so the only place we need to worry about mixing native and Zephyr initializers is with NATIVE_APPLICATION? And that's because there's only a single linker stage? I assume Zephyr is still generating the linker script for that? If so, will that script need to generate a native-compatible initializer list? Which format should it use? Presumably "both", using the old style (count + inits + null) for .ctors and the new style (inits) for .init_array?

If all of my assumptions are correct, then the current patch is nearly correct; it just needs to have the selection between Zephyr and Native intializers corrected so that Native initializers are used with NATIVE_APPLICATION and otherwise we always get Zephyr initializers?

In any case, we'll still need help from someone who understands MetaWare to make sure we can use the same initializers as all other toolchains.

aescolar replied:
> ok, so the only place we need to worry about mixing native and Zephyr initializers is with NATIVE_APPLICATION? And that's because there's only a single linker stage? 

yep & yep

> I assume Zephyr is still generating the linker script for that?

Partly. In this case Zephyr's linker script is an **extension** to the default host linker script (see below).

> If so, will that script need to generate a native-compatible initializer list?

Not necessarily. If for that case we do nothing with .ctors & .init_array's, the host linker script will still generate a fully correct list for the host executable init.
In this NATIVE_APPLICATION case we only need to generate this constructors list if we want to have "our own" to be run in Zephyr's init, wrapped in our own `__zephyr_init_array_start/end`, instead of the host one.
When we generate our own .ctors list, the host one generated by the host linker script will be empty (to be precise, it will not have whatever we have already picked into our list).

Meaning, for NATIVE_APPLICATION, if we do nothing for .ctors in the Zephyr linked script, and do not build in the constructor initialization, the host linker script will still build a correct list, and during the executable init it will be run thru.

-----

Just for reference, this is the relevant part of the host linker script in my case:
```
.preinit_array    :
  {
    PROVIDE_HIDDEN (__preinit_array_start = .);
    KEEP (*(.preinit_array))
    PROVIDE_HIDDEN (__preinit_array_end = .);
  }
  .init_array    :
  {
    PROVIDE_HIDDEN (__init_array_start = .);
    KEEP (*(SORT_BY_INIT_PRIORITY(.init_array.*) SORT_BY_INIT_PRIORITY(.ctors.*)))
    KEEP (*(.init_array EXCLUDE_FILE (*crtbegin.o *crtbegin?.o *crtend.o *crtend?.o ) .ctors))
    PROVIDE_HIDDEN (__init_array_end = .);
  }
  .fini_array    :
  {
    PROVIDE_HIDDEN (__fini_array_start = .);
    KEEP (*(SORT_BY_INIT_PRIORITY(.fini_array.*) SORT_BY_INIT_PRIORITY(.dtors.*)))
    KEEP (*(.fini_array EXCLUDE_FILE (*crtbegin.o *crtbegin?.o *crtend.o *crtend?.o ) .dtors))
    PROVIDE_HIDDEN (__fini_array_end = .);
  }
  .ctors          :
  {
    /* gcc uses crtbegin.o to find the start of
       the constructors, so we make sure it is
       first.  Because this is a wildcard, it
       doesn't matter if the user does not
       actually link against crtbegin.o; the
       linker won't look for a file to match a
       wildcard.  The wildcard also means that it
       doesn't matter which directory crtbegin.o
       is in.  */
    KEEP (*crtbegin.o(.ctors))
    KEEP (*crtbegin?.o(.ctors))
    /* We don't want to include the .ctor section from
       the crtend.o file until after the sorted ctors.
       The .ctor section from the crtend file contains the
       end of ctors marker and it must be last */
    KEEP (*(EXCLUDE_FILE (*crtend.o *crtend?.o ) .ctors))
    KEEP (*(SORT(.ctors.*)))
    KEEP (*(.ctors))
  }
  .dtors          :
  {
    KEEP (*crtbegin.o(.dtors))
    KEEP (*crtbegin?.o(.dtors))
    KEEP (*(EXCLUDE_FILE (*crtend.o *crtend?.o ) .dtors))
    KEEP (*(SORT(.dtors.*)))
    KEEP (*(.dtors))
  }
  ```

ruuddw replied:
> In any case, we'll still need help from someone who understands MetaWare to make sure we can use the same initializers as all other toolchains.

@evgeniy-paltsev is doing some tests to see if the PR works or not. But even if it works 'today', it may break later if MetaWare changes internal structures and/or C/C++ runtime libs.

keith-packard replied:
It's unclear how MetaWare could even break things -- none of the Zephyr initializer code is something the toolchain has influence in, except for the ELF sections assigned to the initializer addresses (currently either `.init_array` or `.ctors`).

In any case, the tests included in this PR should catch any toolchain changes in how initializers are handled by the compiler, so if they work today (and the tests pass), we can be reasonably confident that future changes will not pass unnoticed and Zephyr can adapt as needed. Which isn't really any different from handling toolchain changes anywhere else in the system.

aescolar reviewed (dismissed):
Thanks @keith-packard 
I'm still in favor of the clean up & optimization. But I think the if'ing needs to remain as is (see comments)

ruuddw commented on include/zephyr/arch/arc/v2/linker.ld:
```diff
@@ -142,13 +142,6 @@ SECTIONS {
 #include <snippets-rodata.ld>
 #include <zephyr/linker/kobject-rom.ld>
 
-#if defined(CONFIG_CPP) && !defined(CONFIG_STATIC_INIT_GNU) && defined(__MWDT_LINKER_CMD__)
```
Agreed, my first intuition is that this should remain for MetaWare - but need to check with our toolchain or other people that know more about this.  I expect the _fctors/_ectors are required for the MetaWare runtime library to handle the constructors/destructors. Not sure that can be replaced by other Zephyr initialization code without other consequences/side effects.

keith-packard replied:
Yup, I think that's the best plan -- I *think* this should work correctly even on MetaWare as we're completely replacing the whole constructor path with Zephyr-specific code, effectively bypassing the MetaWare runtime. But, as you say, we need to have the arc people review this before merging.

keith-packard commented on Kconfig.zephyr:
```diff
@@ -434,6 +434,7 @@ config CODING_GUIDELINE_CHECK
 config NATIVE_LIBC
 	bool
 	select FULL_LIBC_SUPPORTED
+	select STATIC_INIT_GNU
```
Thanks -- that's a big help. It sounds like we need a new mode which explicitly constructs a glibc-compatible constructor array, and then only use that when NATIVE_LIBC is enabled.

I'm a bit concerned that we're mixing Zephyr-domain constructors with Linux-domain constructors. I wonder if we should find a way to ensure the Linux ones are always run first?

keith-packard commented:
Ok, I re-jiggered the constructor selection so that the native C library will run all of them when `CONFIG_NATIVE_APPLICATION` is selected. Otherwise, Zephyr runs them all. I think that's right? You mentioned something about CPP being relevant, but I couldn't find where the old code had that condition; maybe it's elsewhere in Kconfig?

aescolar commented on samples/net/wifi/boards/rd_rw612_bga.conf:
```diff
@@ -87,7 +87,7 @@ CONFIG_MBEDTLS_USER_CONFIG_FILE="wpa_supp_els_pkc_mbedtls_config.h"
 CONFIG_MBEDTLS_PSA_CRYPTO_C=y
 CONFIG_MBEDTLS_ENTROPY_C=y
-CONFIG_MBEDTLS_ZEPHYR_ENTROPY=y
+CONFIG_MBEDTLS_ENTROPY_POLL_ZEPHYR=y
```
I think it is best to queue this as a hotfix in a separate PR if it is not already. (CC @valeriosetti )

valeriosetti replied:
This was fixed in https://github.com/zephyrproject-rtos/zephyr/pull/77517.
@jilaypandya thanks for the fix!

aescolar commented:
As mentioned here https://github.com/zephyrproject-rtos/zephyr/pull/74700#discussion_r1729387814, setting DNM until @ruuddw or @evgeniy-paltsev confirm it works with the arc MWDT. Either of you feel free to remove it when you do. Or somebody else if you strongly disagree about waiting for that.

keith-packard commented:
> As mentioned here [#74700 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/74700#discussion_r1729387814), setting DNM until @ruuddw or @evgeniy-paltsev confirm it works with the arc MWDT. Either of you feel free to remove it when you do. Or somebody else if you strongly disagree about waiting for that.

I definitely agree with this plan -- if we're going to clean up this code, we should clean it up everywhere instead of leaving one architecture behind.

evgeniy-paltsev commented:
With the recent changes C++ aplications can't be build with MWDT toolchain:

```
lldac: error: duplicate symbol: __cxa_atexit
>>> defined at arcmwdt-dtr-stubs.c:16 (/verification_workspace/zephyr/arch/arc/arcmwdt/arcmwdt-dtr-stubs.c:16)
>>>            arcmwdt-dtr-stubs.c.obj:(__cxa_atexit) in archive zephyr/libzephyr.a
>>> defined at cpp_dtors.c:41 (/verification_workspace/zephyr/lib/cpp/abi/cpp_dtors.c:41)
>>>            cpp_dtors.c.obj:(.text.__cxa_atexit+0x0) in archive zephyr/libzephyr.a
```

The fix should be trivial, though - to remove __cxa_atexit implementation from arcmwdt-dtr-stubs.c and always use the one from lib/cpp/abi/cpp_dtors.c

keith-packard commented:
> The fix should be trivial, though - to remove __cxa_atexit implementation from arcmwdt-dtr-stubs.c and always use the one from lib/cpp/abi/cpp_dtors.c

This doesn't seem to have anything to do with the current state of this series. Are you suggesting that we add that change to this series? Does the rest of this series work on MWDT, and if so, would you approve it with this additional change?

keith-packard committed:
arch/arc: Remove redundant __cxa_atexit from MWDT stubs

__cxa_atexit was originally defined by the IA64 C++ ABI, but it has been
adopted by other C++ implementations since then.  This function is already
present in lib/cpp/api/cpp_dtors.c, which is built by default when C++
support is enabled, so we don't need another copy in the MWDT-specific
code.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
cpp_dtors.c also defines __dso_handle. Should that also be removed from arcmwdt-dtr-stubs.c?

cfriedt commented:
@evgeniy-paltsev @ruuddw - have you had a chance to run tests on this change for the ARC MataWare toolchain?

M-Moawad commented:
Hi @keith-packard,

I tested this PR using the MWDT toolchain and observed that all C++ tests hang before reaching main(). After some debugging, I discovered the issue is caused by a null pointer being called in the final iteration of the constructor invocation loop during startup:
[init.c#L504](https://github.com/zephyrproject-rtos/zephyr/blob/c7ff57d0e40c75ec7a5da98bd61e5fbcaf4658fc/kernel/init.c#L504).

After checking the map file, I noticed that the list of constructors ends with a NULL entry stored in `__CTORS_END`.

I also reviewed the `__do_global_ctors_aux()` function defined in MWDT and found that its loop includes an additional check to skip over the `__CTORS_END`.

To prevent the null pointer from being invoked, the constructor loop in Zephyr can be modified as follows:

```c
for (fn = __zephyr_init_array_start; fn != __zephyr_init_array_end && *fn; fn++)
```
This added `&& *fn` condition ensures the loop terminates before attempting to call a null pointer.

keith-packard commented:
> I tested this PR using the MWDT toolchain 

Awesome! Thanks for picking this up.

> After checking the map file, I noticed that the list of constructors ends with a NULL entry stored in `__CTORS_END`.

Yup, that's a piece of the old-style constructors array. The other bit is a leading element indicating the number of entries. It looks like the MWDT toolchain doesn't insert that?

Could we fix the generated linker script to simply *not* insert this NULL entry? As Zephyr performs its own interpretation of the data, there's no compatibility issue here. If not, your fix is a simple plan.

> To prevent the null pointer from being invoked, the constructor loop in Zephyr can be modified as follows:
> 
> ```c
> for (fn = __zephyr_init_array_start; fn != __zephyr_init_array_end && *fn; fn++)
> ```
> 
> This added `&& *fn` condition ensures the loop terminates before attempting to call a null pointer.

I'll go ahead and add that. It could be MWDT-toolchain specific, but it seems like a reasonable sanity check anyways.

aescolar commented:
@keith-packard note the checkpatch failure

keith-packard commented:
w00t! Ready for final review.

aescolar reviewed (approved):
Looks good

aescolar commented:
Thanks @M-Moawad I added a DNM label so it does not get merged while you test

aescolar commented:
> I've noticed that the `samples/posix/env` test (details: ....
> Is this timing change expected as part of the updates in this PR? Any insights on why this behavior might have shifted?

This seems quite related to https://github.com/zephyrproject-rtos/zephyr/issues/72858, I tend to recall this test is problematic in SMP targets due to racing of the test and shell.

ruuddw commented:
> This seems quite related to #72858, I tend to recall this test is problematic in SMP targets due to racing of the test and shell.

Thanks for sharing that issue reference, we concluded something similar, the mixed output is not directly related to this PR. Mohamed's question was if someone understands why the timing changes, and if that is 'intended'. For example, is the logical order of Zephyr initialization steps changed and constructors are now called at another moment?

keith-packard commented:
> Thanks for sharing that issue reference, we concluded something similar, the mixed output is not directly related to this PR. Mohamed's question was if someone understands why the timing changes, and if that is 'intended'. For example, is the logical order of Zephyr initialization steps changed and constructors are now called at another moment?

There wasn't any intent to change the order of initialization, nor does this change obviously do anything of that nature. All of the constructors are run at the same time as without this change, which is before `main` starts.

One thing this change *does* do is respect the constructor order that applications define. Before this change, constructors were run in random order on different platforms. Now the linker scripts use the `SORT_BY_INIT_PRIORITY` construct to ensure they are run in the correct order. Maybe that's having an effect on this test?

ruuddw reviewed (approved):
Thanks, as our CI regression issues are not caused directly by this PR I'm ok to merge this now.

aescolar commented:
@ruuddw @M-Moawad , so we remove the DNM? :)

aescolar commented:
@peter-mitsis @andyross a review from either of you is necessary as assignees
